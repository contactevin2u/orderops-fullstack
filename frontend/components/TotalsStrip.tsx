import React from 'react';

export interface Totals {
  subtotal?: number;
  discount?: number;
  delivery_fee?: number;
  return_delivery_fee?: number;
  penalty_fee?: number;
  total?: number;
  paid_amount?: number;
  balance?: number;
}

const currency = new Intl.NumberFormat('ms-MY', {
  style: 'currency',
  currency: 'MYR',
});

function Money({ amount }: { amount: number }) {
  const parts = currency.formatToParts(amount);
  const symbol = parts.find((p) => p.type === 'currency')?.value || '';
  const number = parts.filter((p) => p.type !== 'currency').map((p) => p.value).join('').trim();
  return (
    <span>
      <span>{symbol}</span> {number}
    </span>
  );
}

export default function TotalsStrip({ order }: { order: Totals }) {
  const {
    subtotal = 0,
    discount = 0,
    delivery_fee = 0,
    return_delivery_fee = 0,
    penalty_fee = 0,
    total = 0,
    paid_amount = 0,
    balance,
  } = order || {};

  const toCollect = balance ?? total - paid_amount;

  const rows = [
    { label: 'Subtotal / Jumlah Kecil', value: subtotal },
    { label: 'Discount / Diskaun', value: -Math.abs(discount) },
    { label: 'Delivery / Penghantaran', value: delivery_fee },
    { label: 'Return Delivery / Pengambilan Balik', value: return_delivery_fee },
    { label: 'Penalty / Denda', value: penalty_fee },
    { label: 'Total / Jumlah', value: total, bold: true },
    { label: 'Paid / Dibayar', value: paid_amount },
    { label: 'To Collect / Baki Perlu Dibayar', value: toCollect, bold: true },
  ];

  return (
    <div className="kv">
      {rows.map((r) => (
        <React.Fragment key={r.label}>
          <div>{r.label}</div>
          <div style={r.bold ? { fontWeight: 600, textAlign: 'right' } : { textAlign: 'right' }}>
            <Money amount={r.value} />
          </div>
        </React.Fragment>
      ))}
    </div>
  );
}

