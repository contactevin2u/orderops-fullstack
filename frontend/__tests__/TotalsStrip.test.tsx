import React from 'react';
import { render } from '@testing-library/react';
import TotalsStrip, { Totals } from '@/components/TotalsStrip';

describe('TotalsStrip snapshots', () => {
  const examples: Totals[] = [
    {
      subtotal: 100,
      discount: 0,
      delivery_fee: 10,
      return_delivery_fee: 0,
      penalty_fee: 0,
      total: 110,
      paid_amount: 50,
    },
    {
      subtotal: 200,
      discount: 20,
      delivery_fee: 0,
      return_delivery_fee: 0,
      penalty_fee: 5,
      total: 185,
      paid_amount: 100,
    },
    {
      subtotal: 50,
      discount: 5,
      delivery_fee: 5,
      return_delivery_fee: 5,
      penalty_fee: 10,
      total: 65,
      paid_amount: 65,
    },
  ];

  examples.forEach((order, idx) => {
    it(`renders example ${idx + 1}`, () => {
      const { container } = render(<TotalsStrip order={order} />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
