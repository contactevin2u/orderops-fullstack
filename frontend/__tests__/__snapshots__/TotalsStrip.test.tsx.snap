// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TotalsStrip snapshots > renders example 1 1`] = `
<div
  class="kv"
>
  <div>
    Subtotal / Jumlah Kecil
  </div>
  <div
    style="text-align: right;"
  >
    <span>
      <span>
        RM
      </span>
       
      100.00
    </span>
  </div>
  <div>
    Discount / Diskaun
  </div>
  <div
    style="text-align: right;"
  >
    <span>
      <span>
        RM
      </span>
       
      - 0.00
    </span>
  </div>
  <div>
    Delivery / Penghantaran
  </div>
  <div
    style="text-align: right;"
  >
    <span>
      <span>
        RM
      </span>
       
      10.00
    </span>
  </div>
  <div>
    Return Delivery / Pengambilan Balik
  </div>
  <div
    style="text-align: right;"
  >
    <span>
      <span>
        RM
      </span>
       
      0.00
    </span>
  </div>
  <div>
    Penalty / Denda
  </div>
  <div
    style="text-align: right;"
  >
    <span>
      <span>
        RM
      </span>
       
      0.00
    </span>
  </div>
  <div>
    Total / Jumlah
  </div>
  <div
    style="font-weight: 600; text-align: right;"
  >
    <span>
      <span>
        RM
      </span>
       
      110.00
    </span>
  </div>
  <div>
    Paid / Dibayar
  </div>
  <div
    style="text-align: right;"
  >
    <span>
      <span>
        RM
      </span>
       
      50.00
    </span>
  </div>
  <div>
    To Collect / Baki Perlu Dibayar
  </div>
  <div
    style="font-weight: 600; text-align: right;"
  >
    <span>
      <span>
        RM
      </span>
       
      60.00
    </span>
  </div>
</div>
`;

exports[`TotalsStrip snapshots > renders example 2 1`] = `
<div
  class="kv"
>
  <div>
    Subtotal / Jumlah Kecil
  </div>
  <div
    style="text-align: right;"
  >
    <span>
      <span>
        RM
      </span>
       
      200.00
    </span>
  </div>
  <div>
    Discount / Diskaun
  </div>
  <div
    style="text-align: right;"
  >
    <span>
      <span>
        RM
      </span>
       
      - 20.00
    </span>
  </div>
  <div>
    Delivery / Penghantaran
  </div>
  <div
    style="text-align: right;"
  >
    <span>
      <span>
        RM
      </span>
       
      0.00
    </span>
  </div>
  <div>
    Return Delivery / Pengambilan Balik
  </div>
  <div
    style="text-align: right;"
  >
    <span>
      <span>
        RM
      </span>
       
      0.00
    </span>
  </div>
  <div>
    Penalty / Denda
  </div>
  <div
    style="text-align: right;"
  >
    <span>
      <span>
        RM
      </span>
       
      5.00
    </span>
  </div>
  <div>
    Total / Jumlah
  </div>
  <div
    style="font-weight: 600; text-align: right;"
  >
    <span>
      <span>
        RM
      </span>
       
      185.00
    </span>
  </div>
  <div>
    Paid / Dibayar
  </div>
  <div
    style="text-align: right;"
  >
    <span>
      <span>
        RM
      </span>
       
      100.00
    </span>
  </div>
  <div>
    To Collect / Baki Perlu Dibayar
  </div>
  <div
    style="font-weight: 600; text-align: right;"
  >
    <span>
      <span>
        RM
      </span>
       
      85.00
    </span>
  </div>
</div>
`;

exports[`TotalsStrip snapshots > renders example 3 1`] = `
<div
  class="kv"
>
  <div>
    Subtotal / Jumlah Kecil
  </div>
  <div
    style="text-align: right;"
  >
    <span>
      <span>
        RM
      </span>
       
      50.00
    </span>
  </div>
  <div>
    Discount / Diskaun
  </div>
  <div
    style="text-align: right;"
  >
    <span>
      <span>
        RM
      </span>
       
      - 5.00
    </span>
  </div>
  <div>
    Delivery / Penghantaran
  </div>
  <div
    style="text-align: right;"
  >
    <span>
      <span>
        RM
      </span>
       
      5.00
    </span>
  </div>
  <div>
    Return Delivery / Pengambilan Balik
  </div>
  <div
    style="text-align: right;"
  >
    <span>
      <span>
        RM
      </span>
       
      5.00
    </span>
  </div>
  <div>
    Penalty / Denda
  </div>
  <div
    style="text-align: right;"
  >
    <span>
      <span>
        RM
      </span>
       
      10.00
    </span>
  </div>
  <div>
    Total / Jumlah
  </div>
  <div
    style="font-weight: 600; text-align: right;"
  >
    <span>
      <span>
        RM
      </span>
       
      65.00
    </span>
  </div>
  <div>
    Paid / Dibayar
  </div>
  <div
    style="text-align: right;"
  >
    <span>
      <span>
        RM
      </span>
       
      65.00
    </span>
  </div>
  <div>
    To Collect / Baki Perlu Dibayar
  </div>
  <div
    style="font-weight: 600; text-align: right;"
  >
    <span>
      <span>
        RM
      </span>
       
      0.00
    </span>
  </div>
</div>
`;
